[dynamic shapes] DynamicInts prototype

pianpwk opened the pull request:
Initial prototype for dynamic int inputs, allows users to run with `torch.compile(f)(DynamicInt(4))`, compiling dynamically and using the underlying hint at runtime.

Current behavior:
- Also works in eager (mostly by subclassing int), as scalar input to torch functions, or numpy/math/etc. For example, `x = DynamicInt(3); torch.randn(x); torch.add(y, z, alpha=x); np.arange(x)` all act as if x = 3.
- Behavior for arithmetic ops is to return new DynamicInts rather than static ints; `DynamicInt(3) * 2 = DynamicInt(6)`. This is via SymNode magic methods, but coverage might not be 100% - for example, I had to explicitly override floordiv to avoid int casting. This is not necessarily the case for non-magic method ops (e.g. `math.cos(x)`). The alternative here is to int cast on all operations, but I opted for this for dynamism propagation in non-compiled regions.
- Doesn't ban fullgraph=False; DynamicInt objects might be leaked back to the user, but I guess this is fine, because they can be casted to ints when needed?
- Dynamo only allocates one symbol per DynamicInt; specifying the same DynamicInt for multiple inputs leads to input deduplication, and a guard installed.
- We don't raise on int specialization (in allowlist/maybe_mark_dynamic style) - but an easy change if needed.
- DynamicInts as nn.Module attributes are handled.
- We don't guard on the DynamicInt id, e.g. users can do the following without recompiling (maybe we should guard?)
```python
x = DynamicInt(4)
f(x)
f(1)
f(DynamicInt(3))  # same as f(3)
```


Follow-up work:
- Specifying shape constraints, either at the int-level, e.g.
```python
DynamicInt(64, name="s0", constraints=["s0 % 32 == 0", "s0 <= 1024"]
```
or at the compilation level, e.g. something like
```python
s0 = DynamicInt(64, name="s0")
s1 = DynamicInt(128, name="s1")
with some_compiler_config.dynamic_int_constraints(["s1 == 2*s0", "s0 % 32 == 0"]):
    f(s0, s1)
```
This should subsume the need for specifying derived SymInts?
- SymFloat support - currently it seems backed floats are specialized by the tensorify float pass, and there's no handling in inductor.
- Propagating dynamism in tensor constructors, e.g. `x = DynamicInt(4); torch.randn(x)` could annotate `_dynamo_dynamic_indices`.


Differential Revision: D81698719

cc @ezyang @SherlockNoMad @EikanWang @jgong5 @wenzhe-nrv @voznesenskym @penguinwu @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @jiayisunx @chenyang78 @kadeng @chauhang @amjames @Lucaskabela

williamwen42 commented on torch/_dynamo/source.py:
```diff
@@ -526,6 +526,25 @@ def name(self) -> str:
+    def __post_init__(self) -> None:
+        assert self.base is not None
+
+    def reconstruct(self, codegen: "PyCodegen") -> None:
```
Are you allowing the compiled function to return a DynamicScalar normally? i.e. no graph breaks, but a normal `return` statement.

If so, then I don't think this function needs to be defined. `reconstruct` should only be called for values that are needed outside the compiled region - in both the case of graph break and return.

You might not even need the `self.tx.one_graph` check below then - you can just put an `unimplemented_v2` call here (this should cause dynamo to skip the frame), or assert failure.

williamwen42 commented on torch/_dynamo/source.py:
```diff
@@ -526,6 +526,25 @@ def name(self) -> str:
+    def __post_init__(self) -> None:
+        assert self.base is not None
+
+    def reconstruct(self, codegen: "PyCodegen") -> None:
```
Can you check if this implementation is necessary? You mentioned that this should not escape the compiled region, making me think that we should not ever reconstruct this, but maybe it's possible for this value to be an intermediate value during reconstruction that's not actually returned?

williamwen42 replied:
cc @StrongerXi the source of our discussion on reconstruct vs. what is allowed to escape the compiled region

pianpwk replied:
Ah I guess now simply `codegen(self.base)` works, but that might've also come from other changes around eager handling.

bobrenjc93 replied:
So is this necessary @pianpwk?

pianpwk replied:
I talked to @williamwen42 offline, and I don't think this isn't necessary, but pretty desirable to have, to leak less DynamicInts back to users (at the cost of slightly less dynamism propagation). Added a comment.

ezyang commented:
> Doesn't work in eager yet; x = DynamicInt(3); torch.randn(x); x + 1 won't work

What makes this hard?

ezyang commented:
> Only supports fullgraph=True, as dynamo source handling is responsible for int casting, and we risk leaking a DynamicInt object back to the user as runtime output.

This, in conjunction with eager doesn't work, is confusing to me, actually. If you are forcing fullgraph=True, you don't need to annotate the argument at all; just pass in the dynamic shapes information indirectly to the compiler via the torch.compile (some sort of arg spec, similar to JAX style APIs). The point of an explicit DynamicInt struct is that it CAN propagate around and trigger dynamism at potentially unpredictable spots in the presence of graph breaks. And yes, running in eager should work; if you are unable to trigger `__int__` conversions then the user is obligated to rewrite their code for it!

bobrenjc93 commented:
> And yes, running in eager should work;

+1 this should definitely go hand-in-hand with the compile implementation.

> The point of an explicit DynamicInt struct is that it CAN propagate around and trigger dynamism at potentially unpredictable spots in the presence of graph breaks.

While I think this would be great, I would contend that the biggest benefit of DynamicInt is it finally gives users an affordance to mark dynamism on ints and that it might make sense to take an incremental approach here (eg. doing dynamism propogation in another PR). For example @anijain2305 recently had a rough time marking dynamism of an int property of an nn_module https://github.com/huggingface/transformers/pull/40328/files. @Chillee has mentioned similar woes in the past in his support group. As MSL scales up, I'm sure they'll run into this wall and it'd be nice to have a go-to recommended solution.

> some sort of arg spec, similar to JAX style APIs

Perhaps there is some bikeshedding we need to do, but I personally think something like wrapping your ints in a DynamicInt is more ergonomic than torch.compile kwargs. It's a one line change and as a user, I don't need to learn about the custom DSL we decide on to map the arg spec syntax to the actual spot of dynamism (eg. we use source path for dynamic allowlist and that's not a trivial learning curve). Perhaps I'm misunderstanding the suggestion? If you have an example in mind for how to mark a deeply nested nn.module int property, I'd love to see it!

ezyang commented:
> It's a one line change and as a user, I don't need to learn about the custom DSL we decide on to map the arg spec syntax to the actual spot of dynamism (eg. we use source path for dynamic allowlist and that's not a trivial learning curve). Perhaps I'm misunderstanding the suggestion?

I think it's more tractable if the DSL is (1) solely for input arguments and (2) only works for "structural" types of some sort, so the user can express it in any valid Python syntax and we can map it to the correct Source. For example, if you could only do it for pytrees, I think this is tractable, esp. since that's how export specifies dynamism anyway.

I am not saying we shouldn't do DynamicInt, but we should be doing BOTH of these things, since DynamicInt simulating int is imperfect and there can be cases where it is inconvenient to ensure eager keeps working.

bobrenjc93 commented on test/test_dynamic_shapes.py:
```diff
@@ -1818,6 +1818,95 @@ def test_stride_symnode(self):
+
+        x = DynamicInt(1)
+        z = DynamicInt(3)
+        self.assertEqual(fn(x, x, z), 5)  # guard: x == y
```
instead of hard coding the answer, can you compare with eager?

eg.

fn_c = torch.compile(fn)
self.assertEqual(fn(x, y, z), fn_c(x, y, z))

pianpwk committed:
[dynamic shapes] DynamicInts prototype (#162194)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/162194

Test Plan:
test_dynamic_shapes

Rollback Plan:

Differential Revision: D81698719

pianpwk commented:
ready for another pass
